pkg/boot/linux.go

```diff
@@ -62,7 +63,36 @@ func (li *LinuxImage) String() string {
 	return fmt.Sprintf("LinuxImage(\n  Name: %s\n  Kernel: %s\n  Initrd: %s\n  Cmdline: %s\n)\n", li.Name, stringer(li.Kernel), stringer(li.Initrd), li.Cmdline)
 }
 
-func copyToFile(r io.Reader) (*os.File, error) {
+func copyToFile(r io.Reader, name string) (*os.File, error) {
+	// If source is a regular file in tmpfs, simply re-use that than copy.
+	if _, err := os.Stat(name); err == nil { // Name is the path.
+		if r, _ := mount.IsTmpfs(name); r {
```

Comment:
this is brittle. Nothing guarantees that an *os.File.Name() is actually a valid path. Anybody can call os.NewFile with a random FD and a random string, and that FD can turn out to be anything, and so can that string. I much prefer what I showed you internally, where the caller **explicitly** passes a type that means copying will be skipped, and that type enforces that the file that is being created is a tmpfs file.

Reply:
That is a good point... newFile only simply pass the name to  golang `file` struct -- and `Open` / `OpenFile` on the name would break

Just to understand a bit more before change approach, is that a real use case ? or possible theoretical scenario to rule out for completeness.

Reply:
(ping)

Reply:
I might be missing something, where is *os.File.Name() called?

Reply:
`copyToFile(kernel, stringer(li.Kernel))` == `copyToFile(kernel, kernel.Name())` it's called by the caller.

Reply:
people don't intentionally pass the wrong path. they intentionally pass no path when there isn't one, e.g. in these cases https://github.com/u-root/u-root/search?q=os.NewFile

Reply:
Won't this degrade gracefully? The `os.Stat(name)` will fail is the name is blank and you fall back to the code that copies.

Reply:
THis is why fstat exists

Reply:
Can fstat tell you if the fd is on a tmpfs?

Reply:
I think you can also use `readlink /proc/self/fd/#`, but Linux specific.

Reply:
the file descriptor is not available here, as the kernel and initramfs is passed in via `io.ReaderAt`. 

if original caller passes in a file via `newFile` on a given descriptor, and the `name` given is not legit, I think that is a dead end, as `fstat` system call also depends on a valid path.

but note that this is now within 

```
if fi, err := os.Stat(name); err == nil && fi.Mode().IsRegular() {
   ...
}
```
it won't enter this branch, for the specific case -- the addition of this PR should be a no-op for that specific case, as it will fall back to copyToFile. The question is if that is sth we need to also to avoid copy ? -- I tend think we should copy, as it sounds like original file is probably not in tmpfs

Reply:
I'm not really comfortable with the potential for confusion or the potential for misusing this interface. And I'm not comfortable with that discomfort being ignored.

If stringer(li.Kernel) is e.g. a vfile.File, but the vfile.File.Name() happens to be an actual file path on the real file system, the stat will succeed, and we open and kexec the wrong file here.

If between the time that the file was opened and the file is being used here the file has been changed on the file system, this is different. I'm not comfortable with this solution.

Reply:
If stringer(li.Kernel) is a random file we download from the internet, but (fmt.Stringer).String() happens to return a valid local file path, we'll just open that file and kexec that. That's not cool either.

Reply:
To add another example: Even when File passed in is legit, fits to avoid copy (regular file, on tmpfs), caller may still do sth to sabotage the file to be loaded, e.g the file maybe opened for additional operation (for whatever reason, who knows), and that corrupted it - one temptation is to write code here (in kexec) to prevent that  from happening (note, the operation could just finish before kexec file load call, so it won't confuse with there is ps still editing it)

I don't think we will ever be able to add checks to prevent all kinds of those things. However, we can't  check nothing here neither. There need a balance.

I am happy to incorporate additional checks though, similar to the nature of "source being regular file", "source in tmpfs", etc.

Reply:
IIUC, the latest conversation is about TOCTOU attacks. There are some general techniques around holding onto the same file handle for the duration of your work or comparing inodes to see that multiple stats/opens refer to the same file. https://wiki.sei.cmu.edu/confluence/display/c/FIO45-C.+Avoid+TOCTOU+race+conditions+while+accessing+files

Reply:
My understanding is there is just a lot of complexity in correctly automating the should_copy logic. For simplicity, can this be pushed to the caller who should have a better idea where their image is? The copy could hinge on an i.SkipCopy

Reply:
re: TOCTOU -- that is a good point, re-opening with `name` creates a window that the same file may be swapped. Albeit that window is very small... updated the code a bit, to casting reader back to a `os.File` (a bad assumption?) and re-use the same logic to check tmpfs, etc. That saves a re-opening, should also get rid of other concerns.

re: skip copy -- yea, that is a good one, we could do that -- but at caller side, we still need to check if the directory is on tmpfs or not; but it for sure knows if it operates on a local `os.File`, or other types of file sources

re: interface -- I prefer we reserve this as last option, as this is one-use interface, and I find that a slightly over kill